### PR TITLE
Delay fixes

### PR DESCRIPTION
**Pull request by porres**
the most important commit closes https://github.com/pure-data/pure-data/issues/696

I'm yet to give it a try to https://github.com/pure-data/pure-data/issues/697

### CODE REVIEW

**Comment by millerpuckette** [comment]
I applied first 3 of these - the last one needs more work (I think) to check that delread objects stay consistent with delwrite.

**Comment by porres** [comment]
thanks, the first two commits were actually from another PR (and I screwed up and had it also in here), guess I'll close it https://github.com/pure-data/pure-data/pull/690

**Comment by porres** [comment]
> the last one needs more work (I think) to check that delread objects stay consistent with delwrite.

not sure what you mean, maybe that blocksizes should be of the same size? Anyway, I think this is also to be considered https://github.com/pure-data/pure-data/issues/697

**Comment by porres** [comment]
> thanks, the first two commits were actually from another PR (and I screwed up and had it also in here)

I also have them as a leftovers in this other PR => https://github.com/pure-data/pure-data/pull/694  not sure how it'll work now because they've already been merged... I'll be extra careful not to mess this up again.

**Comment by millerpuckette** [comment]
no worries.  git is deeply confusing, but at least it seems to deal with duplicate commits well.

File: src/d_delay.c [hunk 1]
```diff
@@ -30,6 +30,7 @@ typedef struct _sigdelwrite
     int x_sortno;   /* DSP sort number at which this was last put on chain */
     int x_rsortno;  /* DSP sort # for first delread or write in chain */
     int x_vecsize;  /* vector size for delread~ to use */
+    int x_blocksize;
```
Comment (Spacechild1) [inline src/d_delay.c]: no need to add `x_blocksize`, just use `x_vecsize`

File: src/d_delay.c [hunk 2]
```diff
@@ -123,6 +125,7 @@ static t_int *sigdelwrite_perform(t_int *w)
 
 static void sigdelwrite_dsp(t_sigdelwrite *x, t_signal **sp)
 {
+    x->x_blocksize = sp[0]->s_n;
     dsp_add(sigdelwrite_perform, 3, sp[0]->s_vec, &x->x_cspace, sp[0]->s_n);
     x->x_sortno = ugen_getsortno();
     sigdelwrite_checkvecsize(x, sp[0]->s_n);
```
Comment (Spacechild1) [inline src/d_delay.c]: `sigdelwrite_checkvecsize` will already set `x_vecsize` to the block size.

you would need to make sure that `sigdelwrite_checkvecsize` comes *before* `sigdelwrite_checkvecsize` in `sigdelread_dsp/sigvd_dsp`

File: src/d_delay.c [hunk 3]
```diff
@@ -89,6 +90,7 @@ static void *sigdelwrite_new(t_symbol *s, t_floatarg msec)
     x->x_sortno = 0;
     x->x_vecsize = 0;
```
Comment (Spacechild1) [inline src/d_delay.c]: we should also do `x_rsortno = 0;`

File: src/d_delay.c (hunk 2 above)
Comment (Spacechild1) [reply src/d_delay.c]: > basically, the first object in the DSP chain determines the blocksize

additionally, subsequent calls to `sigdelwrite_checkvecsize` should update `x_vecsize` if `vecsize` is *larger*. This makes sure that we resize the delay line according to the *largest* blocksize.

**Comment by porres** [comment]
well @Spacechild1  , I tried your suggestions, I could compile and test (and it didn't help), but now there are building errors here :/

**Comment by Spacechild1** [comment]
you've missed some details. try this one: https://github.com/pure-data/pure-data/pull/708

**Comment by porres** [comment]
I will, but first let me thank you for all the great work you've been doing :) 

you rock!

**Comment by porres** [comment]
Obviously https://github.com/pure-data/pure-data/pull/708 works like a charm, so I'm just closing this :)

**Comment by porres** [comment]
it seems https://github.com/pure-data/pure-data/pull/708 fixes this, right?

**Pull request closed without merging**